### PR TITLE
Add artifact validation when deploy spring cloud app

### PR DESCRIPTION
**Pull request by Flanker32**
Add artifact validation when deploy spring cloud app

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 1]
```diff
@@ -173,6 +196,52 @@ protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap) {
...
+    private void validateSpringBootApp(String finalJar) throws AzureExecutionException, IOException {
+        final JarFile jarFile = new JarFile(finalJar);
+        final Attributes maniFestAttributes = jarFile.getManifest().getMainAttributes();
+        final String mainClass = maniFestAttributes.getValue(MAIN_CLASS);
+        final String library = maniFestAttributes.getValue(SPRING_BOOT_LIB);
+        if (StringUtils.isEmpty(mainClass)) {
+            throw new AzureExecutionException(MAIN_CLASS_NOT_FOUND);
+        }
+        if (StringUtils.isEmpty(library)) {
+            throw new AzureExecutionException(FAILED_TO_LOAD_DEPENDENCIES);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:208]: FAILED_TO_LOAD_DEPENDENCIES -> Missing Spring-Boot-Lib in ${jarFilePath}!META-INF/MANIFEST.MF, did you use Spring Boot Maven plugin to repackage the target jar?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 2]
```diff
@@ -173,6 +196,52 @@ protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap) {
...
+        }
+        final Map<String, String> dependencies = getSpringAppDependencies(jarFile.entries(), library);
+        if (!dependencies.containsKey(SPRING_BOOT_AUTOCONFIGURE)) {
+            throw new AzureExecutionException(NOT_SPRING_BOOT_PROJECT);
+        }
+        final String springVersion = dependencies.get(SPRING_BOOT_AUTOCONFIGURE);
+        for (String artifact : SPRING_ARTIFACTS) {
+            if (!SpringCloudDependencyManager.isCompatibleVersion(dependencies.get(artifact),
+                                                                  springVersion)) {
+                PluginUtil.showInfoNotificationProject(project, DEPENDENCIES_IS_NOT_UPDATED,
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:218]: provides summary of not compatible versions and missing artifacts

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 3]
```diff
@@ -173,6 +196,52 @@ protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap) {
...
+    private Map<String, String> getSpringAppDependencies(Enumeration<JarEntry> jarEntryEnumeration,
+                                                         String libraryPath) {
+        final List<JarEntry> jarEntries = Collections.list(jarEntryEnumeration);
+        return jarEntries.stream()
+                         .filter(jarEntry -> StringUtils.startsWith(jarEntry.getName(), libraryPath)
+                                 && StringUtils.endsWith(jarEntry.getName(), JAR))
+                         .map(jarEntry -> {
+                             String fileName = StringUtils.strip(jarEntry.getName(), libraryPath);
+                             fileName = StringUtils.stripEnd(fileName, JAR);
+                             final int i = StringUtils.lastIndexOf(fileName, "-");
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:234]: StringUtils.split("ab:cd:ef", ":", 2)    = ["ab", "cd:ef"]

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 4]
```diff
@@ -173,6 +196,52 @@ protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap) {
...
+
+    private Map<String, String> getSpringAppDependencies(Enumeration<JarEntry> jarEntryEnumeration,
+                                                         String libraryPath) {
+        final List<JarEntry> jarEntries = Collections.list(jarEntryEnumeration);
+        return jarEntries.stream()
+                         .filter(jarEntry -> StringUtils.startsWith(jarEntry.getName(), libraryPath)
+                                 && StringUtils.endsWith(jarEntry.getName(), JAR))
+                         .map(jarEntry -> {
+                             String fileName = StringUtils.strip(jarEntry.getName(), libraryPath);
+                             fileName = StringUtils.stripEnd(fileName, JAR);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:233]: misuse of  StringUtils.stripEnd: ```StringUtils.stripEnd("  abcyx", "xyz") = "  abc"```

Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:233]: FilenameUtils.getName

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 5]
```diff
@@ -50,20 +52,40 @@
...
             Arrays.asList(DeploymentResourceStatus.COMPILING,
                           DeploymentResourceStatus.ALLOCATING,
                           DeploymentResourceStatus.UPGRADING);
+    private static final String JAR = ".jar";
+    private static final String MAIN_CLASS = "Main-Class";
+    private static final String SPRING_BOOT_LIB = "Spring-Boot-Lib";
+    private static final String SPRING_BOOT_AUTOCONFIGURE = "spring-boot-autoconfigure";
+    private static final String NOT_SPRING_BOOT_PROJECT = "Project %s is not a spring-boot project.";
+    private static final String FAILED_TO_LOAD_DEPENDENCIES = "Failed to load spring cloud app dependencies";
+    private static final String DEPENDENCIES_IS_NOT_UPDATED = "Azure Spring Cloud dependencies is not updated.";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:84]: dependencies are not updated

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 6]
```diff
@@ -50,20 +52,40 @@
...
+    private static final String MAIN_CLASS_NOT_FOUND =
+            "Main class was not found in current artifact, which is required for spring cloud app.";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:86]: cannot be found in "<jarFile>"

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 7]
```diff
@@ -50,20 +52,40 @@
...
+    private static final String DEPENDENCY_WARNING = "Dependency %s is missing or out of date, you "
+            + "may update the dependencies with `Add Azure Spring Cloud dependency`.";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:88]: Azure -> Add Azure Spring Cloud dependency on project context menu.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:259]: StringUtils.split could set split group number, but seems unable to split from the last appearence

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java [hunk 8]
```diff
@@ -196,41 +200,62 @@ protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap) {
...
-                return;
+            if (!dependencies.containsKey(artifact)) {
+                missingDependencies.add(artifact);
+            } else if (!SpringCloudDependencyManager.isCompatibleVersion(dependencies.get(artifact), springVersion)) {
+                inCompatibleDependencies.put(artifact, dependencies.get(artifact));
             }
         }
+        final String dependencyPrompt = getDependenciesValidationPrompt(
+                missingDependencies, inCompatibleDependencies, springVersion);
+        if (inCompatibleDependencies.size() > 0) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/springcloud/deploy/SpringCloudDeploymentState.java:230]: !inCompatibleDependencies.isEmpty()

**Pull request merged**